File: mosaic-tty/src/jvmMain/java/com/jakewharton/mosaic/tty/TtyOutputStream.java
package com.jakewharton.mosaic.tty;

import java.io.EOFException;
import java.io.IOException;
import java.io.OutputStream;

final class TtyOutputStream extends OutputStream {
	private final Tty tty;

	TtyOutputStream(Tty tty) {
		this.tty = tty;
	}

	@Override
	public void write(int b) throws IOException {
		write(new byte[] { (byte) b }, 0, 1);
	}

	@Override
	public void write(byte[] b, int off, int len) throws IOException {
		while (len > 0) {
			int written = tty.write(b, off, len);
			if (written == -1) {
				throw new EOFException();
			}
			len -= written;
			off += written;
		}
	}
}

File: mosaic-tty/src/jvmMain/kotlin/com/jakewharton/mosaic/tty/TtyOutputStream.kt
package com.jakewharton.mosaic.tty

import java.io.EOFException
import java.io.IOException
import java.io.OutputStream

internal class TtyOutputStream(
	private val tty: Tty,
	private val error: Boolean,
) : OutputStream() {
	private var closed = false

	override fun write(b: Int) {
		val buffer = byteArrayOf(b.toByte())
		write(buffer, 0, 1)
	}

	override fun write(b: ByteArray, off: Int, len: Int) {
		var off = off
		if (!closed) {
			while (true) {
				val written = if (!error) {
					tty.writeOutput(b, off, len)
				} else {
					tty.writeError(b, off, len)
				}
				if (written == -1) break
				off += written
				if (off == len) return
			}
			throw EOFException()
		}
		throw IOException("closed")
	}

	override fun close() {
		closed = true
	}
}
